pilot/pkg/networking/core/v1alpha3/route/route.go

```diff
@@ -998,6 +1001,30 @@ func getEnvoyRouteTypeAndVal(r *route.Route) (envoyRouteType, string) {
 	return iType, iVal
 }
 
+func catchAllMatch(http *networking.HTTPRoute) *networking.HTTPMatchRequest {
+	// A Match is catch all if and only if it has no header/query param match
+	// and URI has a prefix / or regex *.
+	for _, match := range http.Match {
+		catchalluri := false
+		if match.Uri != nil {
+			switch m := match.Uri.MatchType.(type) {
+			case *networking.StringMatch_Prefix:
+				if m.Prefix == "/" {
+					catchalluri = true
+				}
+			case *networking.StringMatch_Regex:
+				if m.Regex == "*" {
+					catchalluri = true
+				}
+			}
+		}
+		if catchalluri && len(match.Headers) == 0 && len(match.QueryParams) == 0 {
```

Comment:
What about other field like Method Authority?

Reply:
Please think about all the fields, otherwise the result maybe not right.

Reply:
As I mentioned, I think we should get it on par with this implementation first (which I think it is doing already) https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/v1alpha3/route/route.go#L993 because that was being used earlier and is being used else where to determine catch all route. If we want to change that logic - should we do it in a separate PR? If this change is causing regression, I think we should fix it as part of this PR. WDYT?

Reply:
Here it is different, say a setting like this, what's the behavior.

```
//   http:
//   - match:
//       uri:
//         prefix: "/"
//       sourceLabels:
//          matchingNoSrc: xxx
//   - match:
//       uri:
//         prefix: "/a"
```

Reply:
I think it should treat it as catchall right because there is a prefix "/" that takes you to the routes given below? I will add a test case to cover this

Reply:
in the above case, no route will be built

Reply:
That could be because of source labels - But not because the catchAll logic is identifying the route as catchAll incorrectly.
I think if we want to do this with out constructing the route, we will have to bring these match related logic in to catchAllMatch so that it rejects in correct catchAll ones. - Is that what u meant?
```
	// Match by source labels/gateway names inside the match condition
	if !sourceMatchHTTP(match, node.WorkloadLabels, gatewayNames) {
		return nil
	}

	// Match by the destination port specified in the match condition
	if match != nil && match.Port != 0 && match.Port != uint32(port) {
	
```

Reply:
I handled this case (in general all catchvalid routes that can not be built) and also applied the same optimization to top level route rules, there also I found we are doing the same thing.

Reply:
It makes me confusing when you provide two functions, why not merging `catchAllMatch` and `catchAllRoute`?

Reply:
I have reverted the `catchAllRoute` change because there were some test failures

Reply:
Debugging it further, I think we should not have this check https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/v1alpha3/route/route.go#L271 - We should process all routes. See https://github.com/istio/istio/blob/master/tests/testdata/config/rule-content-route.yaml routing rule, the tests are passing because the route with no match condition is at the end and the first route is added. If we reverse the order of routes in that file, the tests would fail. 
WDYT?

Reply:
why? translateRoute will take care of the subset right?

Reply:
@hzxuzhonghu I agree with @ramaraochavali that we should first get to parity with https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/v1alpha3/route/route.go#L993 . Otherwise, it would be a breaking change. 

That said, the source match is a very valid question @ramaraochavali . The logic that currently exists in code will handle the source match issue properly (if source did not match, it will return nil and move to the next http rule). But the new logic does not handle the source match. Why cant we change the implementation?

```
out = []
for _, http := range http.Rules
 perRuleOut = []
 for _, match := range http.Match
    perRuleOut = append(perRuleOut, r)
    rType, _ := getEnvoyRouteTypeAndVal(r)
    if rtype is catch all,
       perRuleOut = [r] // overwrite
  out = append(out, perRuleOut)
  if catchAll, break allRoutes
```

Reply:
@rshriram new code also handles source match. First we find if catchall match is there, if is there, we try  `translateRoute` which does source match and source match fails, it returns nil, that point we ignore the catchall route (treat it as invalid) and loop through all matches. I think we should be good with that here.

Reply:
> why? translateRoute will take care of the subset right?

Let us take this case - Should we send both the routes here?  if yes, with the current code, we will skip the second route because of this https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/v1alpha3/route/route.go#L271 as there no matches in the first route. ?
```
   http:
    - route:
      - destination:
          host: c-weighted.extsvc.com
          subset: v1
        weight: 75
    - match:
      - headers:
          version:
            exact: v2
        sourceLabels:
          version: v1
      route:
      - destination:
          host: c-weighted.extsvc.com
          subset: v2
        weight: 25
```

Reply:
We can ignore such types because even if we send, the second rule is irrelavent as order matters here. Second rule when it comes first then only it is useful - Since http routes are ordered list, we can skip after catch all route (i.e. no  matches). So I think we are good with the existing logic there as well.

Reply:
> Why cant we change the implementation?

We can certainly do that - We are trying to avoid `translateRoute` for the routes that we do not send, when catchall routes exist.

Reply:
I have refactored in such a way that we do not unnecessarily call `translateRoute` but still catches the catch all route upfront and skips other routes.  PTAL.